### DISCUSSION TITLE
Create Message with OneDrive Attachment causes error

### DISCUSSION THREAD

**Discussion by dadams39** [microsoftgraph/msgraph-sdk-go#290] [Q&A] [answered]
Found a strange error. Using the following API's:
- [Send Message](https://learn.microsoft.com/en-us/graph/api/user-sendmail?view=graph-rest-1.0&tabs=http)
- [Add Attachment](https://learn.microsoft.com/en-us/graph/api/message-post-attachments?view=graph-rest-1.0&tabs=go)
Go.mod
```java
 github.com/microsoftgraph/msgraph-sdk-go v0.40.0
 github.com/microsoftgraph/msgraph-sdk-go-core v0.28.1
```
Both of the techniques work for an Item Attachment, but for some reason with a OneDrive Attachment, this does not seem to work. Error below:

```bash
 Error Trace:    ..._test.go:407                                                                
                Error:          Received unexpected error:                                                                                                                                         
                                1 error occurred:                                                                                                                                                  
                                        * person@some.onmicrosoft.com:  ErrorInvalidProperty The property '' is required when creating the entity.: error status code received from the API
```                        
Is there any additional documentation on how to send a message with a OneDrive attachment?

**Comment by dadams39** [comment]
This document describes how to add a oneDrive attachment:
https://support.microsoft.com/en-us/office/attach-files-in-outlook-on-the-web-48b8dca1-7a76-43ce-97d1-e1cf73893f55

The intended message described has an attachment that uses the `Upload & Share` option.

**Comment by baywet** [comment, marked as answer]
If you want to attach files from OneDrive/SharePoint (not copying them, but sending a link to them) your application needs to:
1. [upload the file to ODSP](https://learn.microsoft.com/en-us/graph/api/driveitem-put-content?view=graph-rest-1.0&tabs=http) (if not already there).
2. (optional) [create a sharing link](https://learn.microsoft.com/en-us/graph/api/driveitem-createlink?view=graph-rest-1.0&tabs=http) if the recipients are outside of the organization.
3. [create a draft message](https://learn.microsoft.com/en-us/graph/api/user-post-messages?view=graph-rest-1.0&tabs=http)
4. create a [referenceAttachment](https://learn.microsoft.com/en-us/graph/api/resources/referenceattachment?view=graph-rest-1.0) (same as add attachment, except with a different class)
5. [send the message](https://learn.microsoft.com/en-us/graph/api/message-send?view=graph-rest-1.0&tabs=http)

More resources:
- https://stackoverflow.com/questions/64262929/email-with-multiple-attachments-using-graph-api-4-mb/64263205#64263205
- https://learn.microsoft.com/en-us/graph/sdks/large-file-upload?tabs=csharp (unfortunately the Go SDK doesn't support large file upload yet)
